src/Raven.Client/Util/StreamWithTimeout.cs

```diff
@@ -0,0 +1,141 @@
+        {
+            _canBaseStreamTimeoutOnWrite = _stream.CanTimeout && _stream.WriteTimeout < int.MaxValue;
+
+            if (_canBaseStreamTimeoutOnWrite)
+                _writeTimeout = (int?)writeTimeout?.TotalMilliseconds ?? _stream.WriteTimeout;
+            else
+                _writeTimeout = (int)(writeTimeout ?? DefaultWriteTimeout).TotalMilliseconds;
+        }
+
+        public override int ReadTimeout => _readTimeout;
+
+        public override int WriteTimeout => _writeTimeout;
+
+        public override bool CanTimeout => true;
+
+        public override void Flush()
+        {
+            _stream.Flush();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (_canBaseStreamTimeoutOnRead)
+                return _stream.Read(buffer, offset, count);
+
+            return AsyncHelpers.RunSync(() => ReadAsyncWithTimeout(buffer, offset, count, CancellationToken.None));
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (_canBaseStreamTimeoutOnRead)
+                return _stream.ReadAsync(buffer, offset, count, cancellationToken);
+
+            return ReadAsyncWithTimeout(buffer, offset, count, cancellationToken);
+        }
+
+        private Task<int> ReadAsyncWithTimeout(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _cts?.Dispose();
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            _cts.CancelAfter(_readTimeout);
+
+            return _stream.ReadAsync(buffer, offset, count, _cts.Token);
+        }
+
+        
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            return _stream.Seek(offset, origin);
+        }
+
+        public override void SetLength(long value)
+        {
+            _stream.SetLength(value);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            if (_canBaseStreamTimeoutOnWrite)
+            {
+                _stream.Write(buffer, offset, count);
+                return;
+            }
+
+            AsyncHelpers.RunSync(() => WriteAsyncWithTimeout(buffer, offset, count, CancellationToken.None));
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (_canBaseStreamTimeoutOnWrite)
+            {
+                return _stream.WriteAsync(buffer, offset, count, cancellationToken);
+            }
+
+            return WriteAsyncWithTimeout(buffer, offset, count, cancellationToken);
+        }
+
+        private Task WriteAsyncWithTimeout(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            _cts?.Dispose();
+            _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
```

Comment:
It is fine (and fairly common) to do both read & write operations at the same time.
But you are using the same `_cts` on both

Reply:
Done.